cpp/src/fil/infer.cu

```diff
@@ -69,19 +69,19 @@ struct ArgMax {
 
 template <int NITEMS, typename output_type, typename tree_type>
 __device__ __forceinline__ vec<NITEMS, output_type> infer_one_tree(
-  tree_type tree, float* sdata, int cols) {
+  tree_type tree, const float* input, int cols, int rows) {
```

Comment:
necessary to avoid illegal memory access when using L1 cache and `num_rows % NITEMS > 0`

Reply:
If the parameter is not a compile-time constant, the loop cannot be fully unrolled; this can affect performance. It would be better to add a separate template parameter and use `NITEMS` or `rows` as the number of loop iterations based on that.